File: sol/1475/1475.cpp
struct Solution {
    vector<int> finalPrices(vector<int>& prices) {
        int n = prices.size();
        vector<int> res(n, 0), st(1, prices.back());
        res.back() = prices.back();
        for (int i = n - 2; i >= 0; --i){
            while (!st.empty() && st.back() > prices[i]){
                st.pop_back();
            }
            res[i] = prices[i] - (st.empty() ? 0 : st.back());
            st.push_back(prices[i]);
        }
        return res;
    }
};

File: sol/1475/1475.py
class Solution:
    def finalPrices(self, prices: List[int]) -> List[int]:
        n = len(prices)
        res = [0] * n
        res[n - 1] = prices[n - 1]
        st = [prices[n - 1]]
        for i in range(n - 2, -1, -1):
            while len(st) != 0 and st[-1] > prices[i]:
                st.pop()
            res[i] = prices[i] - (0 if len(st) == 0 else st[-1])
            st.append(prices[i])
        return res

